src/common/health.rs

```diff
@@ -0,0 +1,337 @@
+                if peer_id != this_peer_id {
+                    Some(get_consensus_commit(transport_channel_pool, uri))
+                } else {
+                    None
+                }
+            })
+            .collect::<FuturesUnordered<_>>()
+            .inspect_err(|err| log::error!("GetCommitIndex request failed: {err}"))
+            .filter_map(|res| future::ready(res.ok()));
+
+        // Example:
+        //
+        // Total nodes: 2
+        // Required: 2 / 2 = 1
+        //
+        // Total nodes: 3
+        // Required: 3 / 2 = 1
+        //
+        // Total nodes: 4
+        // Required: 4 / 2 = 2
+        //
+        // Total nodes: 5
+        // Required: 5 / 2 = 2
+        let required_commit_indices_count = peer_address_by_id.len() / 2;
+
+        let mut commit_indices: Vec<_> = (&mut requests)
+            .take(required_commit_indices_count)
+            .collect()
+            .await;
+
+        while let Ok(Some(resp)) = time::timeout(Duration::ZERO, requests.next()).await {
+            commit_indices.push(resp);
+        }
+
+        if commit_indices.len() < required_commit_indices_count {
+            log::warn!(
+                "Not enough cluster nodes responded to GetConsensusCommit request: \
+                 required {required_commit_indices_count},
+                 responded {} out of {}",
+                commit_indices.len(),
+                peer_address_by_id.len(),
+            );
+
+            return None;
+        }
+
+        let cluster_commit_index = commit_indices
+            .into_iter()
+            .map(|resp| resp.into_inner().commit)
+            .max()
+            .unwrap_or(0);
+
+        Some(cluster_commit_index as _)
+    }
+
+    fn commit_index(&self) -> u64 {
+        // TODO: Blocking call in async context!?
+        self.consensus_state
+            .persistent
+            .read()
+            .state
+            .hard_state
+            .commit
+    }
+
+    async fn unhealthy_shards(&self) -> HashSet<Shard> {
+        let this_peer_id = self.toc.this_peer_id;
+        let collections = self.toc.all_collections().await;
+
+        let mut unhealthy_shards = HashSet::new();
+
+        for collection in &collections {
+            let state = match self.toc.get_collection(collection).await {
+                Ok(collection) => collection.state().await,
+                Err(_) => continue,
+            };
+
+            for (&shard, info) in state.shards.iter() {
+                let Some(state) = info.replicas.get(&this_peer_id) else {
+                    continue;
+                };
+
+                if is_shard_ready(state) {
+                    continue;
+                }
+
+                unhealthy_shards.insert(Shard::new(collection, shard));
+            }
+        }
+
+        unhealthy_shards
+    }
+
+    fn set_ready(&self) {
+        self.is_ready.store(true, atomic::Ordering::Relaxed);
+        self.is_ready_signal.notify_one();
+    }
```

Comment:
I guess with our current implementation we want to use `notify_waiters` here, because you're unsure how many waiters there are.

What do you think?

Reply:
No problem! Thanks for picking it up.

Reply:
Fixed.